### ISSUE TITLE
Context Menu 'Paste' doesn't work until user pastes with Ctrl+V or uses Context Menu 'Copy'

### ISSUE THREAD

**Issue by niyr** [handsontable/handsontable#3078]
Seeing a repeat occurrence of the issue addressed in #2089 and #2365 when working with HOT 0.20.2 in OSX 10.10 and Chrome 46/47, Firefox 43, Safari 9.

Copying data from Excel or Google Sheets to clipboard and then attempting to paste in a newly initialized HOT instance using the context menu results in a blank cell (doesn't go into edit mode). In Chrome, the table also loses focus after attempting to paste, while on Firefox and Safari the cell remains highlighted. Pasting using Ctrl+V works as expected, and after the user pastes using Ctrl+V the context menu 'Paste' command begins to work as expected.

Using the context menu 'Copy' before 'Paste' works as expected. It's only pasting data that's already on the clipboard using the context menu on a newly initialized HOT instance that appears to be the problem.

Working jsfiddle that reproduces the bug:
http://jsfiddle.net/sauliuslx/ygfnR/341/

Using demo 4 within the HOT docs also produces the same error:
http://docs.handsontable.com/0.20.2/demo-context-menu.html

Can confirm that the zeroclipboard plugin exists and is being loaded correctly through the contextMenuCopyPaste config option.

**Comment by DavidMikeSimon** [comment]
Can confirm this issue is occurring for me as well.

**Comment by DavidMikeSimon** [comment]
Additionally: the problem can reoccur on the same HOT instance with the following steps:

1. Copy data from external source
2. Attempt to paste with context menu in HOT. Fails to paste correctly.
3. Paste with Ctrl+V in HOT, this time it works.
4. Copy another section of data from the external source
5. Attempt to paste with context menu in HOT. **It re-pastes the data that was copied in step 1, not the data from step 4.**

This gives me an inkling of an idea as to what might be causing this, I will investigate further.

**Comment by DavidMikeSimon** [comment]
**TL;DR version:** There is no cross-browser way to directly get the contents of the clipboard. Therefore HOT needs a fallback where it tells the user to press Ctrl+V, which we can and do intercept.

So this whole situation is a mess.

HOT uses ZeroClipboard as a way to copy to the clipboard, but ZeroClipboard has no features for getting data *from* the clipboard at an arbitrary time. No such feature can be added, either, since Flash deliberately makes it impossible for security reasons: http://help.adobe.com/en_US/as3/dev/WS2F6A31B9-1AE6-4b23-9C12-57A33F4F0516.html

There is also a method on document called `execCommand` which can theoretically be used to manually trigger a paste event. However, browser support is extremely spotty. IE supports it in newer versions, Firefox seems to only support it if a manual setting is enabled in the user's config, and Chrome seems to ignore it entirely (except for some kind of workaround that Google Docs uses which I don't understand yet).

It *is* possible to intercept when a user does an explicit paste (e.g. using CTRL+V or selecting "Paste" from the browser's native context menu), and HOT does exactly this in the `copyPaste` plugin, which is why CTRL+V works. Google Docs also relies on this in Firefox; if you try to use the menu item in Google Docs to paste in Firefox, it'll pop up a dialog asking you to press Ctrl+V.

`ContextMenuCopyPaste`'s paste menu item when clicked calls `CopyPasteClass`'s method `triggerPaste` with no arguments. However, this is incorrect; when called with no arguments, `triggerPaste` assumes that the data to be copied is in the hidden textarea that's used to capture manual paste events. This explains the behavior nylr and I were seeing; the ContextMenuCopyPaste menu item simply re-pastes whatever was last intercepted in the hidden textarea from a native copy/paste.

What I'm going to do now is work on a PR that addresses this. My plan is:

1. When `triggerPaste` recieves no argument, it will attempt to use `execCommand` to cause a paste. 

2. We'll know that succeeded if `execCommand` returns true. If it doesn't, `triggerPaste` will pop up a modal asking the user to press Ctrl+V or Cmd+V (depending on their platform).

3. The modal will close on any input event; we can rely on the paste event interceptor that's already implemented to do the hard part.

**Comment by alvarotrigo** [comment]
+1 for the fallback with the message. 
@DavidMikeSimon great explanation!

**Comment by AMBudnik** [comment]
I'm transferring all the improvement suggestion to our Wiki section at: [improvement suggestions](https://github.com/handsontable/handsontable/wiki/Improvement-suggestions)

Please feel free to add any additional suggestions/tips or investigation results. 

Issue is closed. It will be reopened with a developing update.

**Comment by DavidMikeSimon** [comment]
AMBudnik, this is not an improvement suggestion, it is a bug fix or possibly a bug workaround.

**Comment by AMBudnik** [comment]
Hi @DavidMikeSimon 
Could you share a step-by-step scenario how to reproduce the issue you're facing here.
This topic is already very plentiful - I guess I misunderstood the idea.

**Comment by DavidMikeSimon** [comment]
AMBudnik: The long and short of it is that the "Paste" command from the context menu does not and cannot read from the system clipboard. We need to either get rid of the command or implement a workaround. My PR referenced above does the latter.

**Comment by AMBudnik** [comment]
Thanks for feedback @DavidMikeSimon

**Comment by DavidMikeSimon** [comment]
@AMBudnik Anything I should do on my PR to get it okay for merging?

**Comment by AMBudnik** [comment]
@DavidMikeSimon It's always better for developers to get a simple scenario and a demo how to reproduce the issue.

**Comment by DavidMikeSimon** [comment]
@AMBudnik For steps to reproduce the bug, see my comment on this issue at https://github.com/handsontable/handsontable/issues/3078#issuecomment-167850905

**Comment by AMBudnik** [comment]
Oh yes, thanks for pointing that out @DavidMikeSimon

**Comment by swistach** [comment]
Hi folks,
actually this feature never worked properly.  Due to security reasons, reading from the clipboard is not possible.
In v0.33.0/1.12.0 we've decided to show you how you can create your own implementation of `Paste` option in context menu. I hope the tutorial is enough clear.
You can find tutorial here: https://docs.handsontable.com/0.33.0/demo-copy-paste.html#paste-in-context-menu

**Comment by NeoTechni** [comment]
> Hi folks, actually this feature never worked properly. Due to security reasons, reading from the clipboard is not possible. In v0.33.0/1.12.0 we've decided to show you how you can create your own implementation of `Paste` option in context menu. I hope the tutorial is enough clear. You can find tutorial here: https://docs.handsontable.com/0.33.0/demo-copy-paste.html#paste-in-context-menu

we don't need to read from the clipboard, we need to detect paste events. I'm surprised this issue is almost a decade old.
